# Dataset of Nifty for 2018 and 2019 daily numbers - Open,Close, Low, High

In [1]:
import numpy as np
import pandas as pd

In [2]:
nifty50_2018 = pd.read_csv('NIFTY50_2018.csv',index_col=0)

In [3]:
nifty50_2018.head()

Open      High       Low     Close
Date                                               
31 Dec 2018  10913.20  10923.55  10853.20  10862.55
28 Dec 2018  10820.95  10893.60  10817.15  10859.90
27 Dec 2018  10817.90  10834.20  10764.45  10779.80
26 Dec 2018  10635.45  10747.50  10534.55  10729.85
24 Dec 2018  10780.90  10782.30  10649.25  10663.50

In [5]:
nifty50_2018.loc['31 Dec 2018']

Open     10913.20
High     10923.55
Low      10853.20
Close    10862.55
Name: 31 Dec 2018, dtype: float64

In [6]:
nifty50_2019 = pd.read_csv('NIFTY50_2019.csv',index_col=0)

In [7]:
nifty50_2019.head()

Open      High       Low     Close
Date                                               
31 Dec 2019  12247.10  12247.10  12151.80  12168.45
30 Dec 2019  12274.90  12286.45  12213.80  12255.85
27 Dec 2019  12172.90  12258.45  12157.90  12245.80
26 Dec 2019  12211.85  12221.55  12118.85  12126.55
24 Dec 2019  12269.25  12283.70  12202.10  12214.55

In [8]:
#how to club these both datasets
nifty50 = pd.concat([nifty50_2018,nifty50_2019])

In [9]:
nifty50

Open      High       Low     Close
Date                                               
31 Dec 2018  10913.20  10923.55  10853.20  10862.55
28 Dec 2018  10820.95  10893.60  10817.15  10859.90
27 Dec 2018  10817.90  10834.20  10764.45  10779.80
26 Dec 2018  10635.45  10747.50  10534.55  10729.85
24 Dec 2018  10780.90  10782.30  10649.25  10663.50
...               ...       ...       ...       ...
07 Jan 2019  10804.85  10835.95  10750.15  10771.80
04 Jan 2019  10699.70  10741.05  10628.65  10727.35
03 Jan 2019  10796.80  10814.05  10661.25  10672.25
02 Jan 2019  10868.85  10895.35  10735.05  10792.50
01 Jan 2019  10881.70  10923.60  10807.10  10910.10

[491 rows x 4 columns]

In [10]:
nifty_next50 = pd.read_csv('NIFTYNext50_2019.csv',index_col=0)

In [11]:
nifty_next50.head()

Open      High       Low     Close
Date                                               
31 Dec 2019  28495.00  28549.50  28270.25  28307.55
30 Dec 2019  28528.95  28612.95  28406.70  28484.85
27 Dec 2019  28354.50  28500.25  28319.90  28476.80
26 Dec 2019  28409.10  28435.25  28259.75  28280.25
24 Dec 2019  28423.70  28430.40  28318.75  28382.85

In [50]:
# concatenate 2019 indices
nifty_2019 = pd.concat([nifty50_2019,nifty_next50],axis=1,keys=['Nifty50','NiftyNext50'])

In [51]:
nifty_2019.head()

Nifty50                               NiftyNext50            \
                 Open      High       Low     Close        Open      High   
Date                                                                        
31 Dec 2019  12247.10  12247.10  12151.80  12168.45    28495.00  28549.50   
30 Dec 2019  12274.90  12286.45  12213.80  12255.85    28528.95  28612.95   
27 Dec 2019  12172.90  12258.45  12157.90  12245.80    28354.50  28500.25   
26 Dec 2019  12211.85  12221.55  12118.85  12126.55    28409.10  28435.25   
24 Dec 2019  12269.25  12283.70  12202.10  12214.55    28423.70  28430.40   

                                 
                  Low     Close  
Date                             
31 Dec 2019  28270.25  28307.55  
30 Dec 2019  28406.70  28484.85  
27 Dec 2019  28319.90  28476.80  
26 Dec 2019  28259.75  28280.25  
24 Dec 2019  28318.75  28382.85

In [52]:
nifty_2019['Nifty50'].loc['31 Dec 2019']

Open     12247.10
High     12247.10
Low      12151.80
Close    12168.45
Name: 31 Dec 2019, dtype: float64

In [53]:
nifty_2019['Nifty50']['Open']

Date
31 Dec 2019    12247.10
30 Dec 2019    12274.90
27 Dec 2019    12172.90
26 Dec 2019    12211.85
24 Dec 2019    12269.25
                 ...   
07 Jan 2019    10804.85
04 Jan 2019    10699.70
03 Jan 2019    10796.80
02 Jan 2019    10868.85
01 Jan 2019    10881.70
Name: Open, Length: 245, dtype: float64

## 1. In 2019, how many days was the NIFTY50 volatile (high>105% of low)

In [144]:
nifty50.head()


Open      High       Low     Close
Date                                               
31 Dec 2018  10913.20  10923.55  10853.20  10862.55
28 Dec 2018  10820.95  10893.60  10817.15  10859.90
27 Dec 2018  10817.90  10834.20  10764.45  10779.80
26 Dec 2018  10635.45  10747.50  10534.55  10729.85
24 Dec 2018  10780.90  10782.30  10649.25  10663.50

In [152]:
nifty50_2019['Volatile']= nifty50_2019['High']>nifty50_2019['Low']*1.05

In [154]:
nifty50_2019['Volatile'].sum()

1

## 2. In 2019, how many days was the NIFTYNEXT50 volatile (high > 105% of low)

In [108]:
nifty_next50.head()

Open      High       Low     Close
Date                                               
31 Dec 2019  28495.00  28549.50  28270.25  28307.55
30 Dec 2019  28528.95  28612.95  28406.70  28484.85
27 Dec 2019  28354.50  28500.25  28319.90  28476.80
26 Dec 2019  28409.10  28435.25  28259.75  28280.25
24 Dec 2019  28423.70  28430.40  28318.75  28382.85

In [148]:
nifty_next50['Volatile']= nifty_next50['High']>nifty_next50['Low']*1.05

In [149]:
nifty_next50['Volatile'].sum()

1

## 3. In 2019, how many days belonged to the four classes NIFTY50 volatile/non volatile and NIFTYNext50 volatile/non volatile

In [167]:
nifty_vol = pd.concat([nifty50_2019,nifty_next50],axis =1,keys =['n50','nn50'])

In [173]:
nifty_vol[nifty_vol['n50']['Volatile'] & nifty_vol['nn50']['Volatile']]

n50                                          nn50            \
                Open     High      Low    Close Volatile      Open      High   
Date                                                                           
20 Sep 2019  10746.8  11381.9  10691.0  11274.2     True  26026.45  27476.35   

                                          
                  Low     Close Volatile  
Date                                      
20 Sep 2019  25812.05  27406.85     True

In [185]:
nifty_vol[(nifty_vol['n50']['Volatile']== False) | (nifty_vol['nn50']['Volatile']==False)]

n50                                             nn50  \
                 Open      High       Low     Close Volatile      Open   
Date                                                                     
31 Dec 2019  12247.10  12247.10  12151.80  12168.45    False  28495.00   
30 Dec 2019  12274.90  12286.45  12213.80  12255.85    False  28528.95   
27 Dec 2019  12172.90  12258.45  12157.90  12245.80    False  28354.50   
26 Dec 2019  12211.85  12221.55  12118.85  12126.55    False  28409.10   
24 Dec 2019  12269.25  12283.70  12202.10  12214.55    False  28423.70   
...               ...       ...       ...       ...      ...       ...   
07 Jan 2019  10804.85  10835.95  10750.15  10771.80    False  27893.95   
04 Jan 2019  10699.70  10741.05  10628.65  10727.35    False  27707.20   
03 Jan 2019  10796.80  10814.05  10661.25  10672.25    False  27874.05   
02 Jan 2019  10868.85  10895.35  10735.05  10792.50    False  28074.05   
01 Jan 2019  10881.70  10923.60  10807.10  10910.10    False  28243.95   

                                                    
                 High       Low     Close Volatile  
Date                                                
31 Dec 2019  28549.50  28270.25  28307.55    False  
30 Dec 2019  28612.95  28406.70  28484.85    False  
27 Dec 2019  28500.25  28319.90  28476.80    False  
26 Dec 2019  28435.25  28259.75  28280.25    False  
24 Dec 2019  28430.40  28318.75  28382.85    False  
...               ...       ...       ...      ...  
07 Jan 2019  27932.00  27652.95  27706.60    False  
04 Jan 2019  27842.75  27533.90  27719.50    False  
03 Jan 2019  28045.30  27630.20  27674.65    False  
02 Jan 2019  28188.40  27801.75  27876.95    False  
01 Jan 2019  28252.20  27988.60  28111.65    False  

[244 rows x 10 columns]

Since both true case: 1  
      both false case:244  
      Alternate true and false of Nifty50 and Nifty Next 50 is null set

## 4. Compute mean, median, std, var of closing values for each Weekday in NIFTY50

In [77]:
c19 = nifty50_2019['Close']

In [192]:
fdate_index = map(pd.Timestamp,c19.index)
c19_fdate = pd.Series(c19, index= fdate_index)

In [210]:
c19_fdate[c19_fdate.index.dayofweek==6]

2019-10-27    11627.15
Name: Close, dtype: float64

In [206]:
day_c19 = pd.DataFrame({'Weekday':c19_fdate.index.dayofweek, 'Close':c19_fdate})

In [205]:
name ={
    0:'Monday',
    1:'Tuesday',
    2:'Wednesday',
    3:'Thursday',
    4:'Friday',
    5:'Saturday',
    6:'Sunday'
}

def name_fun(i):
    return name.get(i,i)

In [208]:
day_c19['Weekday'] = day_c19['Weekday'].apply(name_fun)

In [209]:
day_c19.groupby('Weekday').describe()

Close                                                             \
          count          mean         std       min         25%        50%   
Weekday                                                                      
Friday     51.0  11435.394118  443.779665  10724.40  11029.3250  11456.900   
Monday     46.0  11463.535870  463.019749  10640.95  11054.8875  11573.475   
Sunday      1.0  11627.150000         NaN  11627.15  11627.1500  11627.150   
Thursday   50.0  11430.373000  470.461726  10672.25  10980.7000  11576.900   
Tuesday    50.0  11444.318000  462.829541  10604.35  10958.0500  11544.150   
Wednesday  47.0  11385.222340  449.850387  10651.80  10974.0500  11440.200   

                                
                 75%       max  
Weekday                         
Friday     11817.225  12271.80  
Monday     11879.775  12262.75  
Sunday     11627.150  11627.15  
Thursday   11843.200  12259.70  
Tuesday    11794.125  12214.55  
Wednesday  11789.175  12221.65

## 5. Compute mean,median,std,var of closing values for each month in NIFTY50

In [119]:
new_c19 = pd.DataFrame({'Month':c19_fdate.index.month, 'Close':c19_fdate})

In [121]:
new_c19

Month     Close
2019-12-31     12  12168.45
2019-12-30     12  12255.85
2019-12-27     12  12245.80
2019-12-26     12  12126.55
2019-12-24     12  12214.55
...           ...       ...
2019-01-07      1  10771.80
2019-01-04      1  10727.35
2019-01-03      1  10672.25
2019-01-02      1  10792.50
2019-01-01      1  10910.10

[245 rows x 2 columns]

In [122]:
res5 = new_c19.groupby('Month')['Close']

In [126]:
res5.describe()

count          mean         std       min         25%        50%  \
Month                                                                     
1       23.0  10809.463043   92.241175  10651.80  10754.7000  10821.600   
2       20.0  10833.840000  119.606867  10604.35  10778.9000  10819.025   
3       19.0  11317.242105  223.401346  10863.50  11113.1250  11354.250   
4       19.0  11666.452632   65.824517  11575.95  11601.2500  11665.950   
5       22.0  11592.022727  279.155921  11148.20  11316.2125  11683.075   
6       19.0  11839.023684  115.116579  11672.15  11756.4750  11841.550   
7       23.0  11523.110870  260.085388  11085.40  11307.6750  11555.900   
8       20.0  10976.500000   97.097806  10741.35  10924.0625  11007.175   
9       19.0  11124.289474  310.891386  10704.80  10846.2750  11003.500   
10      20.0  11490.205000  217.418539  11126.40  11313.8250  11523.300   
11      20.0  11964.132500   84.216057  11840.45  11904.9750  11940.700   
12      21.0  12096.880952  134.226928  11856.80  11994.2000  12086.700   

              75%       max  
Month                        
1      10888.5500  10961.85  
2      10898.3000  11069.40  
3      11472.7250  11623.90  
4      11719.6750  11787.15  
5      11840.1375  11945.90  
6      11910.1250  12088.55  
7      11675.0500  11946.75  
8      11046.5250  11109.65  
9      11457.3250  11600.20  
10     11609.8625  11877.45  
11     12018.4625  12151.15  
12     12221.6500  12271.80

## 6. On the days in which NIFTY50 closed higher than open what was the mean of (close - open) for Nifty Next 50

In [129]:
cond = nifty50_2019['Close'] > nifty50_2019['Open']

In [133]:
(nifty_next50[cond]['Close']-nifty_next50[cond]['Open']).mean()

118.70188679245271

## 7. In 2019, how many days had the day's High lower than previous day's low in N50

In [134]:
succ_high=  nifty50_2019['High'].iloc[1:]

In [135]:
prev_low= nifty50_2019['Low'].iloc[:-1]

In [136]:
np.sum(succ_high.values - prev_low.values < 0)

13

## 8.  In 2019 how many days did the day's close exceed 30 day moving average in NIFTY50

In [138]:
c19_fdate.head()

2019-12-31    12168.45
2019-12-30    12255.85
2019-12-27    12245.80
2019-12-26    12126.55
2019-12-24    12214.55
Name: Close, dtype: float64

In [143]:
np.sum(c19_fdate > c19_fdate.rolling('30d').mean())

100

In [3]:
import pandas as pd
pd.DataFrame.rename?

In [143]:
purchase_1 = pd.Series({'Name': 'Chris',
                        'Item Purchased': 'Dog Food',
                        'Cost': 22.50})
purchase_2 = pd.Series({'Name': 'Kevyn',
                        'Item Purchased': 'Kitty Litter',
                        'Cost': 2.50})
purchase_3 = pd.Series({'Name': 'Vinod',
                        'Item Purchased': 'Bird Seed',
                        'Cost': 5.00})

df = pd.DataFrame([purchase_1, purchase_2, purchase_3], index=['Store 1', 'Store 1', 'Store 2'])

In [124]:
df.index.name='Location'

In [125]:
df

Name Item Purchased  Cost
Location                            
Store 1   Chris       Dog Food  22.5
Store 1   Kevyn   Kitty Litter   2.5
Store 2   Vinod      Bird Seed   5.0

In [107]:
df.loc['Store 3']= ['Laurette','Fish Food',3.20]

In [108]:
df

Name Item Purchased  Cost
Location                               
Store 1      Chris       Dog Food  22.5
Store 1      Kevyn   Kitty Litter   2.5
Store 2      Vinod      Bird Seed   5.0
Store 3   Laurette      Fish Food   3.2

In [127]:
df.set_index([df.index,'Name'],inplace=True)

In [136]:
df.drop(columns='Laurette',inplace=True)

In [121]:
df = df.append(pd.Series(['Kitty Food',3.00],index = ['Item Purchased','Cost'],name =('Store 2','Kevyn')))

In [122]:
df

Item Purchased  Cost
Location Name                      
Store 1  Chris       Dog Food  22.5
         Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0
         Kevyn     Kitty Food   3.0

In [128]:
df.index

MultiIndex([('Store 1', 'Chris'),
            ('Store 1', 'Kevyn'),
            ('Store 2', 'Vinod')],
           names=['Location', 'Name'])

In [134]:
df.loc[('Store 3','Laurette')] = pd.Series({'Item Purchased':'Fish Food','Cost':3.20})

In [140]:
df.drop('Store 3',inplace=True)

In [141]:
df

Item Purchased  Cost
Location Name                      
Store 1  Chris       Dog Food  22.5
         Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0

In [144]:
df.sort_index()

Name Item Purchased  Cost
Store 1  Chris       Dog Food  22.5
Store 1  Kevyn   Kitty Litter   2.5
Store 2  Vinod      Bird Seed   5.0